### PR TITLE
1051 - Upload video to kaltura using cron jobs

### LINKED ISSUE #989: Upload video content to Kaltura instead of Wordpress

### LINKED ISSUE #1051: Upload video to kaltura using cron jobs

### PR DESCRIPTION
**Pull request by EldadZZipori**
## Description
In this PR we want to change how videos are being uploaded by the server. We want the process to be transparent for the user, with no changes at all to the user UI and ideally with no client side interaction at all.
image

Use #990 to implement the back-end kaltura api as it is already (mostly) implemented there.

*Update*: As of May 2022, the functionality of this PR no longer involves setting up 301 redirects for the old videos

## Changes
Backend/Kaltura:
- Adds backend functionality to upload videos to Kaltura using Kaltura API
- Adds a custom Wordpress Action "upload_videos_to_kaltura" that triggers the upload to Kaltura

Frontend:
- Re-fetches the video URL stored in the node if the existing URL becomes unavailable while the frontend is open (e.g. if the video was uploaded and deleted from local server)

Other:
- Adds a section to the README about setting up Kaltura
- Fixes detection of TAPESTRY_USE_DEV_MODE
- Defines TAPESTRY_USE_DEV_MODE in wp-config in Wordpress installation script for CI and Docker

*Note*: Changes made in this PR do not include setting up the cron job, nor setting up 301 redirects

## Change Details
- Organized the uploaded videos using [Kaltura Categories](https://developer.kaltura.com/api-docs/Secure_Control_and_Govern/Content-Categories-Management.html), which have a tree-like hierarchy. We are saving all videos inside the parent category 'Tapestry' at the root level. Within 'Tapestry', videos are further categorized by date of upload.
- Added a custom Wordpress Option "tapestry_kaltura_upload_in_progress" to track whether an upload is ongoing, to avoid starting a new upload while upload is in progress

## Testing
In order to test this PR, first make sure to follow the README to ensure Kaltura works.
In addition, go to Wordpress settings->media and make sure "Organize my uploads into month- and year-based folders" is unchecked.

The PR itself only creates the Wordpress action, so to test it, you can create a temporary endpoint in endpoints.php that triggers the action.

Insert the following entry in the `REST_API_ENDPOINTS` array:
```
    'UPLOAD_VIDEOS_TO_KALTURA' => (object) [
        'ROUTE' => '/kaltura/upload_videos',
        'ARGUMENTS' => [
            'methods' => $REST_API_GET_METHOD,
            'callback' => 'uploadVideosToKaltura',
        ],
    ],
```

Then, still in endpoints.php, add the following function below
```
function uploadVideosToKaltura($request) {
    do_action('upload_videos_to_kaltura');
}
```

You can then trigger the video upload by going to http://localhost:8888/wp-json/tapestry-tool/v1/kaltura/upload_videos. Once the request completes (takes quite a while), the videos should all be uploaded to Kaltura.

Also, you can see the value of the "tapestry_kaltura_upload_in_progress" Option in phpMyAdmin.

~~To test the functionality of this PR use the link http://localhost:8000/wp-json/tapestry-tool/v1/kaltura (with postmaster) to initial the uploading sequence.~~

## NOTES:
~~Video data includes video URL and thumbnail information.~~
~~301 redirect is in case someone is accessing the old URL (did not refresh their tapestry yet etc.)~~
~~Label 301 redirect in .htaccess by date for cleaning up redirects in retrospect~~
- There is a separate task to pull thumbnail with a button

## Useful links - 
Cron jobs - https://kinsta.com/knowledgebase/wordpress-cron-job/
301 Redirect - https://themeisle.com/blog/301-redirect-wordpress/
Kaltura API documentation for developers – [uploading videos](https://developer.kaltura.com/workflows/Ingest_and_Upload_Media/Uploading_Media_Files), [Kaltura Categories](https://developer.kaltura.com/api-docs/service/category)

## Issue Linkage
Closes #989; Closes #1051

## PR Dependency
Depends on: #1163
Depends on #1167

### CODE REVIEW

**Comment by EldadZZipori** [comment]
After testing this PR whether it works or not please re-assign to me so I can make the function a task and not an endpoint

File: tapestry.php [hunk 1]
```diff
@@ -316,3 +316,85 @@ function tapestry_tool_log_event()
...
     wp_die();
 }
+
+
+// Kaltura
+add_action('upload_videos_to_kalture', 'upload_videos_to_kalture');
+function upload_videos_to_kaltura()
+{
+    if (LOAD_KALTURA) {
+        $upload_folder = getcwd()."/wp-content/uploads";
```
Comment (wynnset) [inline tapestry.php:326]: It would be more reliable to use the Wordpress [wp_upload_dir](https://developer.wordpress.org/reference/functions/wp_upload_dir/) function to get the upload dir

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "require": {
+        "kaltura/api-client-library": "^16.14",
+        "spatie/async": "^1.5"
```
Comment (wynnset) [inline composer.json:4]: Is the async library needed anymore?

File: tapestry.php [hunk 3]
```diff
@@ -316,3 +316,85 @@ function tapestry_tool_log_event()
...
+                    $node->set($typeData);
+                    $node->save();
+
+                    wp_delete_file($upload_folder."/".$original_link);
+                    break;
+                }
+            }
+        }
+
+        error_log("Background Kaltura uplaod complete.");
```
Comment (wynnset) [inline tapestry.php:398]: Typo here. Also if you are going to write to the log, please also include the original link and/or the node ID as well so it's more useful.

File: templates/vue/README.md [hunk 4]
```diff
@@ -125,3 +125,4 @@ Again you can view a more comprehensive report on your browser using:
 ```
 npm run report:open
 ```
+
```
Comment (wynnset) [inline templates/vue/README.md:128]: Remove empty line

File: templates/vue/cypress/integration/lightbox/activity.spec.js [hunk 5]
```diff
@@ -397,7 +397,6 @@ describe("Activity", () => {
       cy.getByTestId("question-answer-text-multi-0").click({ force: true })
       cy.submitModal()
       cy.openLightbox(node.id)
-      cy.contains(/Change answers/i).click()
```
Comment (wynnset) [inline templates/vue/cypress/integration/lightbox/activity.spec.js:400]: These seem unrelated to this PR.

File: templates/vue/src/components/Lightbox/media/ActivityMedia/index.vue [hunk 6]
```diff
@@ -1,10 +1,7 @@
 <template>
   <div ref="activity" class="activity-media">
     <h1 v-if="showTitle" class="media-title">{{ node.title }}</h1>
-    <completion-screen
-      v-if="state === 'completion-screen'"
-      :question="activeQuestion"
-    >
+    <completion-screen v-if="showCompletionScreen" :question="activeQuestion">
```
Comment (wynnset) [inline templates/vue/src/components/Lightbox/media/ActivityMedia/index.vue:4]: Changes in this file seem unrelated to this PR.

File: templates/vue/src/components/Lightbox/media/TapestryMedia.vue [hunk 7]
```diff
@@ -56,16 +56,22 @@
       @load="handleLoad"
     ></wp-post-media>
     <activity-media
-      v-if="node.mediaType === 'activity' || node.mediaType === 'answer'"
```
Comment (wynnset) [inline templates/vue/src/components/Lightbox/media/TapestryMedia.vue:59]: Changes in this file seem unrelated to this PR.

File: templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue [hunk 8]
```diff
@@ -126,6 +127,18 @@ export default {
         currentTime,
       })
     },
+    handleError() {
+      // Error code 4 - The associated resource or media provider object has been found to be unsuitable.
+      if (this.$refs.video.error.code === 4) {
+        if (
```
Comment (wynnset) [inline templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue:133]: Should we automatically try and fetch this node from the server (once only) and use its updated media URL before showing the refresh option?

**Review by wynnset — CHANGES_REQUESTED** [review]
Great job overall! Just some polishing and it's good to go

File: tapestry.php (hunk 3 above)
Comment (EldadZZipori) [reply tapestry.php:398]: Every cron job affects several nodes and videos. I thought to might be useful to print this in case something goes wrong. I did not add a date because wp does that already. What do you think?

File: templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue (hunk 8 above)
Comment (EldadZZipori) [reply templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue:133]: Done

File: README.md [hunk 9]
```diff
@@ -46,3 +46,38 @@ It's strongly recommended you complete this to get link previews working for the
...
+1. In your wordpress root directory, run `composer install`. If successful, you will be able to see a `vendor` folder containing `kaltura`.
+2. Edit your Wordpress `wp-config.php` file and add the following lines right above the comment "That's all, stop editing! Happy publishing":
+    ```
+    define('KALTURA_ADMIN_SECRET', '');
+    define('KALTURA_PARTNER_ID', '');
+    define('KALTURA_SERVICE_URL', '');
+    ```
+The Kaltura Admininstrator Secret and Partner ID can be found by going to your Kaltura Settings > Integration tab in the Kaltura admin. The service URL is simply the main domain where your Kaltura videos are hosted on.
+
+3. On wordpress admin, open Settings->Media and make sure `Organize my uploads into month- and year-based folders` is un-checked
```
[the remark this answers is not in the document]
Comment (EldadZZipori) [reply README.md:63]: What we can do is change the folder for tapestry in general to use our folder. I looked at the backend and it seems we did not implement our own upload, but we can change our upload folder on wpData to a folder exclusively for tapestry plugins. What do you think?

[the remark this answers is not in the document]
Comment (wynnset) [reply README.md:63]: Yes, that sounds great!

File: endpoints.php [hunk 10]
```diff
@@ -308,6 +317,13 @@
             'permission_callback' => 'TapestryPermissions::putTapestrySettings',
         ],
     ],
+    'KALTURA_TEST' => (object) [
```
Comment (wynnset) [inline endpoints.php:320]: Please do not leave TEST code in here. If this is not for testing, please change its name to something more meaningful like `TRANSFER_VIDEOS_TO_KALTURA` or something. 

For security purposes, it might also be a good idea to add a hashed url parameter that acts as a secret for running this operation. You can check that using a custom `permission_callback`.

File: endpoints.php [hunk 11]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
         return new WP_Error($e->getCode(), $e->getMessage(), $e->getStatus());
     }
     
+}
+
+function upload_videos_to_kaltura()
+{
+    if (LOAD_KALTURA) {
```
Comment (wynnset) [inline endpoints.php:1514]: Before allowing this to be run, we should check and make sure this operation is not running already. You can save an option named `tapestry_kaltura_upload_in_progress` for instance using the Wordpress [update_option()](https://developer.wordpress.org/reference/functions/update_option/) function (set it to true at the beginning and to false after done) and retrieve it using get_option() (making sure it's not running, otherwise stopping automatically). We should make sure to handle timeouts as well so it's not "stuck" in upload mode forever.

File: endpoints.php [hunk 12]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
...
+                            "partnerId" => $kaltura_data->partnerId,
+                        );
+
+                        $node->set($typeData);
+                        $node->save();
+
+                        wp_delete_file($file_obj->file_path);
+                        
+                        
+
```
Comment (wynnset) [inline endpoints.php:1563]: Remove multiple empty lines.

File: endpoints.php [hunk 13]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
...
+                        wp_delete_file($file_obj->file_path);
+                        
+                        
+                       
+                    }
+                }
+            }
+        }
+
+
```
Comment (wynnset) [inline endpoints.php:1569]: Remove multiple empty lines.

File: endpoints.php [hunk 14]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
...
+        foreach ($tapestries as $value) {
+            $tapestry = new Tapestry($value->ID);
+
+            foreach ($tapestry->getNodeIds() as $node_id) {
+                $node = new TapestryNode($value->ID, $node_id);
+                $nodeMeta = $node->getMeta();
+
+                if ($nodeMeta->mediaType == "video") {
+                    $media_url = $node->getTypeData()->mediaURL;
+                    if (strpos($media_url, "/wp-content/uploads/")) {
```
Comment (wynnset) [inline endpoints.php:1533]: It's generally frowned upon to hardcode the upload directory. Use wp_upload_dir function instead.

File: endpoints.php [hunk 15]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
...
+                        $offset = strlen("/wp-content/uploads/");
```
Comment (wynnset) [inline endpoints.php:1534]: This variable is never used.

File: endpoints.php [hunk 16]
```diff
@@ -1491,4 +1507,65 @@ function getQuestionHasAnswers($request)
...
+                        $kaltura_data = null;
+                        try {
+                            $kaltura_data = $kalturaApi->uploadKalturaVideo($file_obj, $current_date);
+                        } catch (Exception $e){
+                            error_log("Unable to upload video - ".$file_obj->file_path." to kaltura, ".$e);
+                            return;
+                        }
+
+                        $typeData = $node->getTypeData();
+                        $typeData->mediaURL = $kaltura_data->dataUrl."?.mp4";
```
Comment (wynnset) [inline endpoints.php:1551]: I'm confused by the notation `...dataUrl."?.mp4"` here. Is the question mark supposed to be there? What does it do?

**Review by wynnset — CHANGES_REQUESTED** [review]
Almost there! I still have to test this, but found some issues with the code that needs to be addressed.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (wynnset) [reply README.md:63]: @EldadZZipori why did you mark this conversation as resolved? Does this not need to be removed now?!

**Comment by wynnset** [comment]
@EldadZZipori please fix the tests - sorry I broke them

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (zifgu) [reply README.md]: Is this part of the README outdated? I could edit it.

File: endpoints.php (hunk 11 above)
Comment (zifgu) [reply endpoints.php]: I've added some code to do this, although it currently does not handle timeouts as far as I know.

File: tapestry.php [hunk 17]
```diff
@@ -132,7 +137,7 @@ function tapestry_enqueue_vue_app()
     global $post;
     if ('tapestry' == get_post_type($post) && !post_password_required($post)) {
         global $TAPESTRY_VERSION_NUMBER;
-        global $TAPESTRY_USE_DEV_MODE;
+        $TAPESTRY_USE_DEV_MODE = defined('TAPESTRY_USE_DEV_MODE') && !empty(TAPESTRY_USE_DEV_MODE);
 
         $use_dev = $TAPESTRY_USE_DEV_MODE || isset($_GET['debug']);
```
Comment (wynnset) [inline tapestry.php:142]: I would merge these into a single line since `$TAPESTRY_USE_DEV_MODE` is not used anywhere else and it doesn't add much clarity here to have it in 2 lines.

(not linted)
```suggestion
        $use_dev = (defined('TAPESTRY_USE_DEV_MODE') && !empty(TAPESTRY_USE_DEV_MODE)) || isset($_GET['debug']);
```

Comment (zifgu) [reply tapestry.php]: Got it. This has now been changed.

File: classes/class.kaltura-api.php [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+            if ($videoCategoryIndex) {
+                $videoCategory = (false !== $videoCategoryIndex ? $categories->objects[$videoCategoryIndex] : null);
+            } else {
+                $category = new Category();
+                $category->parentId = $parentCategory->id;
+                $category->name = $categoryName;
+                $kAdminClient = $this->getKClient(SessionType::ADMIN);
+                $videoCategory = $kAdminClient->category->add($category);
+            }
```
[the remark this answers is not in the document]
Comment (zifgu) [reply classes/class.kaltura-api.php]: I'm not 100% sure of the details here, but I think this code is preparing the Kaltura Categories the video will be added to, for organization purposes. I have updated the PR description with more details about Categories and added some comments. I'm happy to write a bit more if needed.

**Comment by legendword** [comment]
Video uploading seems to work as expected. The new video URL is accessible and the `tapestry_kaltura_upload_in_progress` option is being correctly set. Closing and reopening the node successfully fetches the new URL.

However, I noticed that if I do not close the modal when the upload is done, and at this point if I skip forward to an uncached portion of the video, the video will simply fail to play and the URL is not being properly updated. Seems like the error code is 2 in this case. (Failure on initial loading of the video still works as expected, so I'm assuming this different error code is due to loading more data from an already established source.)

image

File: templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue [hunk 19]
```diff
@@ -147,6 +150,24 @@ export default {
         currentTime,
       })
     },
+    async handleError() {
+      // Error code 4 - The associated resource or media provider object has been found to be unsuitable.
+      if (this.$refs.video.error.code === 4) {
```
Comment (legendword) [inline templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue:155]: Maybe we should remove this `if` and perform the node update check on any type of video loading error?

**Review by legendword — CHANGES_REQUESTED** [review]
Video uploading seems to work as expected. The new video URL is accessible and the `tapestry_kaltura_upload_in_progress` option is being correctly set. Closing and reopening the node successfully fetches the new URL.

However, I noticed that if I do not close the modal when the upload is done, and at this point if I skip forward to an uncached portion of the video, the video will simply fail to play and the URL is not being properly updated. Seems like the error code is 2 in this case. (Failure on initial loading of the video still works as expected, so I'm assuming this different error code is due to loading more data from an already established source.)

image

File: templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue (hunk 19 above)
Comment (zifgu) [reply templates/vue/src/components/Lightbox/media/VideoMedia/UrlVideoMedia.vue:155]: Good catch. This solution sounds good, I can change it shortly.

File: classes/class.kaltura-api.php [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+            /**
+             * Uploading Video Steps:
+             * 1. Create upload token
+             * 2. Upload the file data
+             * 3. Create Kaltura Media Entry
+             * 4. Attach the video.
+             */
```
[the remark this answers is not in the document]
Comment (wynnset) [reply classes/class.kaltura-api.php]: Sorry marking this as unresolved because it's still not done.

File: classes/class.kaltura-api.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (wynnset) [reply classes/class.kaltura-api.php]: Please add the explanation of what's going on here and why as comments in the code.

**Review by wynnset — CHANGES_REQUESTED** [review]

**Review by wynnset — APPROVED** [review]

**Review by legendword — APPROVED** [review]
URL is properly replaced while viewing the video now.

**Comment by zifgu** [comment]
@wynnset Starting the discussion of how the upload will be triggered, now that it is working. Should we create an endpoint for the Kaltura upload?

**Comment by zifgu** [comment]
The upload can be initiated using #1169 , so marking this as ready for merge.

**Comment by wynnset** [comment]
@zifgu is this ready? If so, please mark it as "ready for review" and assign to me.

**Comment by zifgu** [comment]
> @zifgu is this ready? If so, please mark it as "ready for review" and assign to me.

Yes, sorry. The only changes made since William's review have been merges to keep the branch up to date, so I will mark it as ready.

**Review by wynnset — APPROVED** [review]

**Comment by wynnset** [comment]
Merged into #1169
